팔로우 관련 기능 추가

EverJun2 opened the pull request:
1. 팔로윙 / 언팔로윙 기능 추가 
**(POST)               /api/users/follow/{opponentId}    -> 팔로윙
(DELETE)          /api/users/unfollow/{opponentId} -> 언팔로윙**

2. 팔로워 조회 기능 추가
**(GET)    /api/users/{currentUserId}/follower**  

3. 특정 팔로워 확인 기능 추가
**(GET)   /api/users/{currentUserId}/follower/{opponentId}**

4. 팔로윙 조회 기능 추가
**(GET)   /api/users/{currentUserId}/following**

5. 특정 팔로윙 확인 기능 추가
**(GET)  /{currentUserId}/following/{opponentId}**

6. 팔로워 수 조회 기능 추가
**(GET) api/users/{currentUserId}/follower/count**

7. 팔로윙 수 조회 기능 추가
**(GET) api/users/{currentUserId}/following/count**

EverJun2 committed:
feat : Follow entity 작성

EverJun2 committed:
feat : FollowDTO 작성

EverJun2 committed:
refactor : Follow Entity 생성 안되는 문제 해결

EverJun2 committed:
feat : FollowController 생성

EverJun2 committed:
feat : FollowRepository 생성

EverJun2 committed:
feat : FollowServiceImpl 생성

EverJun2 committed:
fix : 특정 팔로워/팔로윙 , 팔로워 수 , 팔로윙 수 기능 버그 수정

EverJun2 committed:
refactor : 팔로윙 / 언팔로윙 api 주소 변경 (로그인 후 테스트 필요)

whrjsgml0000 commented on src/main/java/club/inq/team1/entity/Follow.java:
```diff
@@ -0,0 +1,41 @@
+    @Column(name = "alarm", nullable = false)
+    private Boolean alarm = false;
+
+    public Follow(User followerId, User followeeId) {
```
이 부분이 필요하면 Lombok 어노테이션에 AllArgsConstructor 이용하면 편할거에요

whrjsgml0000 commented on src/main/java/club/inq/team1/service/impl/FollowServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+    private final UserRepository userRepository;
+
+    //팔로우
+    public ResponseEntity<?> follow(Long currentUserId, Long opponentId) {
```
서비스에서 ResponseEntity 를 반환 값으로 가질 필요는 없어요. 비지니스 로직을 처리한 다음에 컨트롤러에서 필요할 만한 데이터만 보내주는 것으로 바꾸고 컨트롤러에서 ResponseEntity를 이용해서 View에 보내주는 방식으로 바꾸는게 좋을 것 같아요.

whrjsgml0000 replied:
그리고 ResponseEntity 에 제네릭으로 와일드카드를 이용하는건 위험할 것 같아요

whrjsgml0000 commented on src/main/java/club/inq/team1/service/impl/FollowServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+import org.springframework.stereotype.Service;
+
+@Log4j2
+@RequiredArgsConstructor
```
RequiredArgsConstructor 를 이용해서 깔끔하게 생성자 주입하는건 좋아보여요! 나중에 리팩토링하는 과정에서 다른 파일들도 저렇게 바꿀까 생각중이에요

whrjsgml0000 commented on src/main/java/club/inq/team1/service/impl/FollowServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+        Follow follow = new Follow(follower, followee);  // `Follow` 엔티티 생성
+        followRepository.save(follow);  // DB에 저장
+
+        return new ResponseEntity<>("팔로우 성공", HttpStatus.CREATED);
```
컨트롤러에서 ResponseEntity를 사용할 때 `ResponseEntity.status().body()` 이런 빌드 패턴을 이용해서 자주 사용해요.

whrjsgml0000 commented on src/main/java/club/inq/team1/service/impl/FollowServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+        User follower = userRepository.findById(currentUserId)
+            .orElseThrow(() -> new IllegalArgumentException("현재 사용자가 존재하지 않습니다."));
+        User followee = userRepository.findById(opponentId)
+            .orElseThrow(() -> new IllegalArgumentException("팔로우 할 사용자가 존재하지 않습니다."));
```
예외 상황에 대한 처리를 꼼꼼하게 잘 해주신 것 같아요!

whrjsgml0000 commented on src/main/java/club/inq/team1/service/impl/FollowServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+    public ResponseEntity<?> unfollow(Long currentUserId, Long opponentId) {
+        // 팔로우 관계를 찾고 삭제
+        Optional<User> currentUser = userRepository.findById(currentUserId);
+        Optional<User> opponentUser = userRepository.findById(opponentId);
```
이 부분도 위에 한 것 처럼 orElseThrow를 이용하면 좋을 것 같아요

whrjsgml0000 commented on src/main/java/club/inq/team1/repository/FollowRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+    List<Follow> findFolloweesByFollowerId(User currentUser);
+    List<Follow> findFollowersByFolloweeId(User currentUser);
+    List<Follow> findByFollowerIdAndFolloweeId(User opponentUser, User currentUser);
```
이 부분은 return 값으로 Optional<Follow>로 바꾸는게 좋아보여요.

whrjsgml0000 commented on src/main/java/club/inq/team1/controller/FollowController.java:
```diff
@@ -0,0 +1,118 @@
+@Tag(name = "FollowController", description = "팔로윙 관련 API 컨트롤러")
+public class FollowController {
+    private final FollowServiceImpl followService;
+    private final UserServiceImpl userServiceImpl;
```
FollowService 인터페이스를 먼저 만든 뒤에 그걸 상속받는 followServiceImpl을 만들고 인터페이스의 기능을 구현하는 방식으로 만드는 게 좋아요. 
UserService는 현재 그렇게 만들어져 있기 때문에 의존성 주입을 할 때 `private final UserService userService` 라고 해도 실제로 작동할 때는 Service 어노테이션이 지정된 UserServiceImpl이 주입됩니다.

whrjsgml0000 commented on src/main/java/club/inq/team1/controller/FollowController.java:
```diff
@@ -0,0 +1,118 @@
+    @GetMapping("/{currentUserId}/following/count")
+    private ResponseEntity<?> countFollowee(@PathVariable Long currentUserId){
+        List<Follow> followees = followService.findAllFollowees(currentUserId);
+        return new ResponseEntity<>(followees.size(), HttpStatus.OK);
```
followees.size() 대신 FollowRepository에 `countByFollowerId(User userId)` 를 이용해서 받아오는 방법은 어떨까요? 이건 그냥 의견이에요!

whrjsgml0000 commented on src/main/java/club/inq/team1/service/impl/FollowServiceImpl.java:
```diff
@@ -0,0 +1,102 @@
+        Optional<User> currentUser = userRepository.findById(currentUserId);
+        if (currentUser.isPresent()) {
+            return followRepository.findFolloweesByFollowerId(currentUser.get());  // currentUser.get()로 User 객체를 전달
+        } else {
```
위의 조건문에서 return 이 있기 때문에 else가 없더라도 로직은 동일하게 작동해요! else는 제거하는게 코드보기도 더 좋을 것 같아요

whrjsgml0000 reviewed (commented):
전체적으로 주석을 달아주셔서 보기 편했어요! 앞으로도 이렇게 작성해주시면 너무 좋을 것 같아요. 저도 주석 꼼꼼히 만들어서 PR 하도록 할게요! 고생 많으셨어요

EverJun2 committed:
refactor : Follow class All args Constructor 애노테이션 적용 및 follower, followee로 변경

EverJun2 committed:
refactor : Follow class All args Constructor 애노테이션 적용 취소

whrjsgml0000 reviewed (changes requested):
조금만 손보면 될 것 같아요. 제가 Follow 클래스에서 AllArgsConstructor 쓰라고 한건 alarm을 생각안하고 있어서 그랬던거고 NonNull 어노테이션이랑 RequiredArgsConstructor 이용하면 되긴 할거에요. 굳이 안하셔도 됩니다 ㅎㅎ

EverJun2 committed:
refactor : Follow.class @RequiredArgsConstructor 애노테이션 적용

EverJun2 committed:
refactor : 팔로워/팔로윙 수 로직을 service 클래스로 이동 및 컨트롤러 리턴 방식 수정

EverJun2 committed:
refactor : 팔로우/언팔로우 controller 리턴 값 수정

EverJun2 committed:
refactor : 팔로우/언팔로우 service 리팩토링

EverJun2 committed:
refactor : 특정 팔로윙/팔로워 확인 service 리팩토링

EverJun2 committed:
refactor : FollowServiceImpl checkUsers() 메소드 생성으로 중복 코드 제거 등 전체 코드 리팩토링

EverJun2 committed:
refactor : FollowController ResponseEntity 반환 구조 리팩토링

EverJun2 committed:
feat : FollowService 인터페이스 생성

EverJun2 committed:
refactor : FollowService 인터페이스 생성에 따른 코드 수정

whrjsgml0000 committed:
Merge branch 'develop' of https://github.com/InQ-InQ-InQ-InQ-InQ/2025-Project-Team1-Winter-BE into feature/follow

whrjsgml0000 committed:
refactor : 코드 수정

whrjsgml0000 committed:
feat : 데이터 전송 구조 변환

whrjsgml0000 committed:
refactor : 팔로윙 수, 팔로워 수 파악 기능을 repository 레벨에서 해결하도록 변경

whrjsgml0000 committed:
refactor : 구체적인 리턴값 파악

whrjsgml0000 committed:
feat : interface를 이용한 프로젝션 구현

+ Follower에 대해서도 구현 필요
+ 프론트엔드에서 이용하기 편하도록 데이터의 평탄화 필요
+ 페이징을 이용해서 조금씩 가져오도록 수정 필요

whrjsgml0000 committed:
feat : 팔로워 리스트를 받을때 interface projection 으로 필요한 정보만 받도록 수정

whrjsgml0000 committed:
feat : 팔로워 팔로윙 리스트를 가져올 때 page를 이용해 조금씩 가져올 수 있도록 수정

whrjsgml0000 committed:
refactor : 코드 정리

whrjsgml0000 committed:
feat : mapper 생성 및 응답 데이터 평탄화
